transport/src/main/java/io/netty/channel/CoalescingBufferQueue.java

```diff
@@ -0,0 +1,225 @@
+     */
+    public void add(ByteBuf buf) {
+        ObjectUtil.checkNotNull(buf, "buf");
+        if (readableBytes > Integer.MAX_VALUE - buf.readableBytes()) {
+            throw new IllegalStateException("buffer queue length overflow: " + readableBytes
+                    + " + " + buf.readableBytes());
+        }
+        bufAndListenerPairs.add(buf);
+        readableBytes += buf.readableBytes();
+    }
+
+    /**
+     * Add a buffer to the end of the queue and associate a promise with it that should be completed when
+     * all the buffers bytes have been consumed from the queue and written.
+     * @param buf to add to the tail of the queue
+     * @param promise to complete when all the bytes have been consumed and written, can be void.
+     */
+    public void add(ByteBuf buf, ChannelPromise promise) {
+        // buffers are added before promises so that we naturally 'consume' the entire buffer during removal
+        // before we complete it's promise.
+        ObjectUtil.checkNotNull(promise, "promise");
+        if (promise.isVoid()) {
+            add(buf);
+        } else {
+            add(buf, new ChannelPromiseNotifier(promise));
+        }
+    }
+
+    /**
+     * Add a buffer to the end of the queue and associate a listener with it that should be completed when
+     * all the buffers  bytes have been consumed from the queue and written.
+     * @param buf to add to the tail of the queue
+     * @param listener to notify when all the bytes have been consumed and written.
+     */
+    public void add(ByteBuf buf, ChannelFutureListener listener) {
+        // buffers are added before promises so that we naturally 'consume' the entire buffer during removal
+        // before we complete it's promise.
+        ObjectUtil.checkNotNull(buf, "buf");
+        ObjectUtil.checkNotNull(listener, "listener");
+        if (readableBytes > Integer.MAX_VALUE - buf.readableBytes()) {
+            throw new IllegalStateException("buffer queue length overflow: " + readableBytes
+                    + " + " + buf.readableBytes());
+        }
+        bufAndListenerPairs.add(buf);
+        bufAndListenerPairs.add(listener);
+        readableBytes += buf.readableBytes();
+    }
+
+
+    /**
+     * Remove a {@link ByteBuf} from the queue with the specified number of bytes. Any added buffer who's bytes are
+     * fully consumed during removal will have it's promise completed when the passed aggregate {@link ChannelPromise}
+     * completes.
+     *
+     * @param bytes the maximum number of readable bytes in the returned {@link ByteBuf}, if {@code bytes} is greater
+     *              than {@link #readableBytes} then a buffer of length {@link #readableBytes} is returned.
+     * @param aggregatePromise used to aggregate the promises and listeners for the constituent buffers.
+     * @return a {@link ByteBuf} composed of the enqueued buffers.
+     */
+    public ByteBuf remove(int bytes, ChannelPromise aggregatePromise) {
```

Comment:
This was collapsed by the time I hit submit...

I think other places that need to combine multiple promises into 1 use something like `ChannelPromiseAggregator`. Ignoring extra object allocation of `Entry` it is almost like you want the signature of remove to be:

```java
public Entry<ByteBuf, ChannelPromise> remove(int bytes) {
}
```

the returned `ChannelPromise` can be a single entry from the queue (if sizes are appropriate), or a `ChannelPromiseAggregator` (or something like it).

If we move toward a linked list approach you may not even have to worry about an extra allocation, because we could use  `Entry<ByteBuf, ChannelPromise>` as our elements in the list.